### PR TITLE
refactor update public jobs query

### PR DESCRIPTION
**Pull request by yakkomajuri**
## Changes

*Please describe.*  
*If this affects the frontend, include screenshots.*  

`addPublicJobIfNotExists` is a slow query. This should speed things up a bit.

## How did you test this code?

*Please describe.*

Added tests

### CODE REVIEW

**Review by tiina303 — COMMENTED** [review]
Why is this better/faster? I feel like we've been moving towards the direction of using less rather than more transactions

**Comment by yakkomajuri** [comment]
@tiina303 The Postgres JSONB operations have been very slow. Doing this on the Node side should be faster + allows us to be more selective with issuing update queries.

No need to worry about this transaction though, this is a query that is run a negligible amount of times when it comes to transaction load on the server, but the time it takes is not negligible and can be problematic.

Screenshot 2022-02-15 at 18 27 21

Notice how we get a few runs an hour, and that's across all of our plugin servers with all their workers on the Cloud cluster. Also notice how long it takes.

File: plugin-server/src/utils/db/db.ts [hunk 1]
```diff
@@ -1585,4 +1585,36 @@ export class DB {
...
+                publicJobs = { ...publicJobs, [jobName]: jobPayloadJson }
+
+                await this.postgresQuery(
+                    'UPDATE posthog_plugin SET public_jobs = $1 WHERE id = $2',
+                    [JSON.stringify(publicJobs), pluginId],
+                    'updatePublicJob',
+                    client
+                )
+            }
+        })
```
Comment (mariusandra) [inline plugin-server/src/utils/db/db.ts:1618]: A transaction doesn't guarantee these rows run sequentially with nothing else happening on the table in between. It'll look ugly, but to guarantee nothing slips through the cracks, you should do the `if` and object merging in postgres.

Comment (yakkomajuri) [reply plugin-server/src/utils/db/db.ts:1618]: It does because of this: `FOR UPDATE`

Comment (yakkomajuri) [reply plugin-server/src/utils/db/db.ts:1618]: As in _a_ transaction by itself does not. But _this_ transaction does, because of that clause

Comment (mariusandra) [reply plugin-server/src/utils/db/db.ts:1618]: Ooh, TIL. Thanks! 🎉

**Review by tiina303 — APPROVED** [review]

**Pull request merged**